django/chat/utils.py

```diff
@@ -357,253 +288,90 @@ def title_chat(chat_id, force_title=True):
-        )
-        reduce_template_fr = (
-            "Les résumés suivants ont été générés à partir d'un document:\n\n"
-            "{docs}\n\n"
-            f"Prenez-les et réécrivez-le en un résumé final. {length_prompt_fr}. Renvoyez-le au format markdown.\n"
-            "Réponse utile:"
+            custom_prompt
+            + "\n\n"
+            + _("The original document is below, enclosed in triple quotes:")
+            + "\n'''\n{docs}\n'''"
         )
-        reduce_template = (
-            reduce_template_fr if target_language == "fr" else reduce_template_en
-        )
-        reduce_prompt = PromptTemplate.from_template(reduce_template)
-
-        map_chain = LLMChain(llm=llm, prompt=map_prompt)
-
-        # Run chain
-        reduce_chain = LLMChain(llm=llm, prompt=reduce_prompt)
-
-        # Takes a list of documents, combines them into a single string, and passes this to an LLMChain
-        combine_documents_chain = StuffDocumentsChain(
-            llm_chain=reduce_chain, document_variable_name="docs"
-        )
-
-        # Combines and iteratively reduces the mapped documents
-        reduce_documents_chain = ReduceDocumentsChain(
-            # This is final chain that is called.
-            combine_documents_chain=combine_documents_chain,
-            # If documents exceed context for `StuffDocumentsChain`
-            collapse_documents_chain=combine_documents_chain,
-            # The maximum number of tokens to group documents into.
-            token_max=15000,
-        )
-
-        # Combining documents by mapping a chain over them, then combining results
-        map_reduce_chain = MapReduceDocumentsChain(
-            # Map chain
-            llm_chain=map_chain,
-            # Reduce chain
-            reduce_documents_chain=reduce_documents_chain,
-            # The variable name in the llm_chain to put the documents in
-            document_variable_name="docs",
-            # Return the results of the map steps in the output
-            return_intermediate_steps=False,
-        )
-
-        # Split the text into chunks
-        text_splitter = RecursiveCharacterTextSplitter.from_tiktoken_encoder(
-            chunk_size=12000, chunk_overlap=100
-        )
-
-        docs = [Document(page_content=text, metadata={"source": "userinput"})]
-
-        split_docs = text_splitter.split_documents(docs)
-
-        async def run_map_reduce():
-
-            # Run the chain asynchronously and return the result when it's done
-            return await map_reduce_chain.arun(split_docs)
+    # Tree summarizer prompt requires certain variables
+    # Note that we aren't passing in a query here, so the query will be empty
+    length_prompt_template = length_prompt_template.replace(
+        "{docs}", "{context_str}{query_str}"
+    )
+    template = PromptTemplate(length_prompt_template, prompt_type=PromptType.SUMMARY)
 
-        response = asyncio.run(run_map_reduce())
+    response = llm.tree_summarize(
```

Comment:
Simpler than with previous Langchain implementation. This will only make multiple LLM calls if the input tokens exceed the model's maximum (in case of GPT-4o, 128k). This very rarely happens, so you get the response sooner. It is also cheaper without the additional "reduce" calls to summarize the chunk summaries. I haven't noticed a drop in quality vs. the map reduce summaries but also haven't tested much.